### PR TITLE
[BC-breaking] C++ API parity: at::Tensor::grad

### PR DESCRIPTION
**Pull request by pbelevich**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26332 C++ API parity: at::Tensor::requires_grad_
* #26251 C++ API parity: at::Tensor::detach
* #26217 C++ API parity: at::Tensor::version
* #26216 C++ API parity: at::Tensor::output_nr
* #26186 C++ API parity: at::Tensor::is_leaf
* **#26150 [BC-breaking] C++ API parity: at::Tensor::grad**

Pull Request resolved: https://github.com/pytorch/pytorch/pull/26150

Differential Revision: [D17427579](https://our.internmc.facebook.com/intern/diff/D17427579)

### CODE REVIEW

**Comment by yf225** [comment]
@pbelevich Thanks for the tests! Curious did you resolve why there is a difference in behavior between Python and C++ for the following code?
```python
# Python

In [1]: import torch

In [2]: x = torch.ones(1, dtype=float, requires_grad=True)

In [3]: x.backward()

In [4]: x.grad
Out[4]: tensor([1.], dtype=torch.float64)
```
```cpp
// C++
auto tensor = torch::ones(1, at::TensorOptions().dtype(at::kFloat).requires_grad(true));
tensor.backward();
std::cout << tensor.grad() << std::endl;
// [ Tensor (undefined) ]
```
From our earlier discussion, Python `tensor.backward()` is defined in `torch/tensor.py`, which has a different implementation than `Variable::backward()`, and it would be ideal to let the C++ implementation also call `torch::autograd::backward` to have feature parity with Python. Curious do you see any roadblocks when making the change?

**Review by yf225 — COMMENTED** [review]
Thanks a lot for fixing `Tensor::backward()`! I believe we will need a BC-breaking note in the PR description to explain what previously working C++ API use cases would break after this PR.

**Review by yf225 — APPROVED** [review]
We can land this PR after BC-breaking note is added

**Comment by suo** [comment]
This appears to have broken the XLA build cc @ailzhang

**Comment by ailzhang** [comment]
Thanks @suo! https://github.com/pytorch/xla/pull/1065 will be merged soon to fix it! :D

**Comment by gchanan** [comment]
where is the actual note saying what is BC breaking?

**Comment by gchanan** [comment]
ok, it looks like the BC issue is that for non-1-element tensors you no longer create the grad_output to pass through backwards.
